src/ann_solo/utils.py

```diff
@@ -1,104 +1,338 @@
+    features["group"] = _get_ssm_groups(ssms, min_group_size) if grouped else 0
+    dataset = mokapot.dataset.LinearPsmDataset(
+        features,
+        target_column="is_target",
+        spectrum_columns="index",
+        peptide_column="sequence",
+        group_column="group",
+    )
+    if grouped:
+        logging.debug(
+            "Partitioned %d SSMs into %d groups",
+            len(ssms),
+            features["group"].nunique(),
+        )
+    # Define the mokapot model.
+    #   - Choice between a random forest, linear SVM, or no semi-supervised
+    #     learning.
+    #   - Features are preprocessed by standardizing them, removing
+    #     zero-variance features, and removing highly correlated features.
+    #   - We perform minimal tuning of TODO hyperparameters.
+    if model is None:
+        logging.debug("Calculate q-values based on the cosine similarity")
+        # Calculate q-values based on the cosine similarity.
+        confidences = dataset.assign_confidence(features["cosine"], True)
+    else:
+        logging.debug(
+            "Train semi-supervised %s model and score SSMs", model.upper()
+        )
+        scaler = make_pipeline(
+            StandardScaler(),
+            VarianceThreshold(),
+            CorrelationThreshold(0.95),
+        )
+        if model == "svm":
+            clf = mokapot.model.PercolatorModel(scaler, train_fdr=fdr)
+        elif model == "rf":
+            clf = mokapot.Model(
+                GridSearchCV(
+                    RandomForestClassifier(random_state=1),
+                    param_grid={
+                        "max_depth": [3, 5, 7, 9, None],
+                        "class_weight": [
+                            {0: neg, 1: pos}
+                            for neg in (0.1, 1, 10)
+                            for pos in (0.1, 1, 10)
+                        ],
+                    },
+                    refit=False,
+                    cv=3,
+                    n_jobs=-1,
+                ),
+                scaler,
+                train_fdr=fdr,
+            )
         else:
-            break
+            raise ValueError(
+                "Unknown semi-supervised machine learning model given"
+            )
+        # Train the mokapot model and combine the SSMs for all groups.
+        confidences, _ = mokapot.brew(dataset, clf, fdr, folds=5)
+    ssm_scores = pd.concat(
+        [
+            confidences.group_confidence_estimates[group].psms[
+                confidences.group_confidence_estimates[group].psms[
+                    'mokapot q-value'] <= fdr]
+            for group in confidences.groups
+        ],
+        ignore_index=True,
+    )
+    # Assign the scores to the SSM objects.
+    for i, score, q in zip(
+        ssm_scores["index"],
+        ssm_scores["mokapot score"],
+        ssm_scores["mokapot q-value"],
+    ):
+        ssms[i].search_engine_score = score
+        ssms[i].q = q
+        yield ssms[i]
```

Comment:
There's not really an advantage of using an iterator here, because we've already consumed all SSMs previously anyway. I suggest reverting to the `return ...`. Alternatively, the docstring and function type need to be updated accordingly.

Reply:
The return statement retruns the whole list of PSM matches found by ANN-SoLo after assigning the new score and the q-value for all positive targets. What we need is just the true target entries matching our FDR to be returned as our best matches (same logic you had previously in the old ANN-SoLo version). Otherwise, you always get the full list as the input query.

This statement: confidences.group_confidence_estimates[group].psms , returns all True targets with their corresponding score and q-value. So I changed that to filter for entries above the FDR threshold in the concatenation then return those matches later on in the above loop.

Reply:
Indeed, I think it could be useful to separate rescoring from FDR filtering. So first return the full list from this function, and then in the calling function that implements the cascade strategy do the filtering and submitting the remaining spectra to the second round. That way we can uncouple unrelated functionality between (re)scoring and FDR filtering.

So your changes are correct (thx!), I'd just put them in a different location in the code.

Reply:
Would also work. Was just following the same logic we had in the current version of ANN-SoLo to have correct output as I was running the whole software.

Reply:
Wanted to get some results on our large test set, but CalCua is down for 4 days this week :/

Reply:
Yes, makes sense. This indeed follows how ANN-SoLo was doing it originally, but I've since realized that that's not always the best solution. But this is not documented, so impossible to know that in advance of course.